### PR TITLE
[WIP] Pointy stax

### PR DESCRIPTION
**Pull request by j-towns**
Some models are quite hard (maybe impossible?) to express in terms of serial and parallel composition of layers. See for example [pixelcnn++](https://arxiv.org/abs/1701.05517), whose topology looks like [this](https://drive.google.com/file/d/1V2wu9IkzLaA0OrlqEU9Htd-sOsjRrS_Z/view?usp=sharing).

This pull request implements an API extension to stax, partly inspired by [Keras's functional API](https://keras.io/getting-started/functional-api-guide/), which allows one to define more sophisticated network structures.

Usage at the moment looks like this:
```python
import jax.experimental.stax as stax

# One layer dense net with relu nonlinearity
def nn_fun(inputs):
  return stax.Relu(stax.Dense(out_dim=4)(inputs))

init_fun, apply_fun = stax.pointy_to_stax_layer(nn_fun)

inputs = np.array([[1., 2., 3.],
                   [4., 5., 6.]])

# nn_fun(inputs)  <- This will raise a ValueError!!

_, params = init_fun(inputs.shape)
out = apply_fun(params, inputs)
```

I'm very open to feedback on this. In particular, I'm not sure if it's the right decision to over-ride the existing stax layers rather than defining new names for the pointy versions. Keras does over-ride, so this might be convenient for new JAX users who are moving over. But on the other hand it might cause confusion. Also the name `pointy_to_stax_layer` is pretty clunky so maybe we can rename that.

TODO:
- [ ] Handle layers with tuple inputs
- [ ] Docstring for `pointy_to_stax_layer`
- [ ] Make the params data-structure more human interpretable
- [ ] Tests?

### CODE REVIEW

**Comment by levskaya** [comment]
Hi Jamie - do you have a slightly more complex pointy use-case example showing how this might be used for pointy-style variable re-use?  I'm quite interested in exploring this approach.  I've achieved similar through crazy hacks via closures at the moment, but I like the idea of using jax's tracing engine for this problem.

**Comment by jekbradbury** [comment]
I could be missing something, but I think this gets at the distinction I brought up recently in the context of JAX neural net libraries between pointiness in what we might call "layer builders," pointiness in "layer instances," and pointiness in "symbolic tensors."

Like the Keras functional API, `pointy_to_stax_layer` provides pointiness in symbolic tensors that's absent from plain stax. If we assume that `+` is overloaded to create a `stax.parallel` and a `stax.sum`, something like this should work for a basic example of a non-linear topology:
```python
def residual_dense(x):
  return x + stax.Dense(out_dim=4)(x)
init_fn, apply_fn = pointy_to_stax_layer(residual_conv)
```
Here `x` is essentially a symbolic tensor: `pointy_to_stax_layer` will trace it once and build a stax AST, and further calls (including at runtime or in the `jit` tracer) will call the `+` inside `stax.sum`, not the one in `residual_conv`. This kind of pointiness is the AST-level equivalent of creating a stax layer with a custom `apply_fn` (where symbolic tensors are already pointy).

It also has pointiness in uninitialized layer constructor objects that I'm calling layer builders (but here it's not materially different from original stax):
```python
# with `pointy_to_stax_layer`:
def double_dense(x):
  dense = stax.Dense(out_dim=4)
  return dense(dense(x))
init_fn, apply_fn = pointy_to_stax_layer(double_dense)
# with plain stax:
dense = stax.Dense(out_dim=4)
init_fn, apply_fn = stax.serial(dense, dense)
```
Here `dense` is not an initialized layer but simply a recipe for a layer; in either of the two snippets above the two dense layers will share the same constructor arguments but will be initialized separately (the returned `init_fn` will call the `init_fn` of `dense` twice).

What Anselm is asking for is pointiness in layer objects, which would provide the ability to alias initialized layers ("share parameters"/"tie weights"). One way to do this is TensorFlow-style `get_variable`, but we'll pretend that never happened :slightly_smiling_face:. Another is a vaguely Lasagne-style API, something approximately like this where the default is pointiness in layer instances, but pointiness in symbolic tensors is available by accessing a layer instance attribute:
```python
def shared_params_dense(x):
  dense1 = Dense(x, out_dim=4)
  dense2 = Dense(dense1.out, out_dim=4, w=dense1.w)
  return dense2.out
```
Alternatively the default can be pointiness in symbolic tensors while pointiness in layer instances is available if needed:
```python
def shared_params_dense(x):
  y = Dense(x, out_dim=4)
  dense1 = y.layer
  return Dense(y, out_dim=4, w=dense1.w)
```
And yet one more way was introduced in Torch's nngraph library, where pointiness in layer instances and symbolic tensors play parallel roles to pointiness in layer builders and symbolic tensors in this PR:
```python
def shared_params_dense(x):
  dense = Dense(in_dim=4, out_dim=4) # in_dim is needed because this line actually initializes the layer
  return dense(dense(x))
```
In kind of a dual of the symbolic tensor situation, this kind of pointiness is the AST-level equivalent of creating a stax layer with a custom `init_fn` (where layer/parameter instances are already pointy):
```python
def double_with_shared_params(layer):
  def init_fn():
    layer_init, _ = layer
    layer_instance = layer_init()
    return (layer_instance, layer_instance)
  def apply_fn(x, params):
    _, layer_apply = layer
    l1, l2 = params
    return layer_apply(layer_apply(x, p1), p2)
```

**Comment by j-towns** [comment]
Nice! That's quite a lot to take in, you've got a lot further than I had thinking about this. I hadn't thought about 'weight sharing' at all, but obviously that's a pretty common requirement.

> It also has pointiness in uninitialized layer constructor objects that I'm calling layer builders (but here it's not materially different from original stax)

I'm probably misunderstanding something, are you saying all things which can be implemented in pointy stax (the current implementation in this pr), can be implemented in original stax? I'm not sure that's true, although I haven't thought too hard about it. I couldn't see how to implement the pixelcnn++ model that I linked a picture of above in original stax.

In case anyone's interested, the code for the original pixelcnn++ model is [here](https://github.com/openai/pixel-cnn/blob/master/pixel_cnn_pp/model.py), and my JAX/pointy stax port is [here](https://gist.github.com/j-towns/23acfd9f1b1a2239539917190b7c01be).

**Comment by j-towns** [comment]
I'm wondering whether pointy stax should maybe become the default way to define a NN. It seems more flexible and user-friendly than using `parallel` and `serial` to me. It's conceptually simpler imo because all it's doing is removing weights from the function definition.

We should allow for weight sharing/using the same layer in more than one place, and I think the most straightforward way to do that is to have the layer functions return a layer instance, rather than a layer builder, so you can build an RNN like this:
```python
def rnn(inputs):
    dense_r = Dense(4)
    recurrent_layer = lambda inputs: Relu(dense_r(inputs))
    for _ in range(sequence_length):
        inputs = recurrent_layer(inputs)
    return inputs
```
This will mean you have to do more layer instantiations, which in a network with many layers of identical type (but different weights) you'd end up with a little more code, but I think that's a relatively rare use case.

It'll be straightforward to implement the above. At the moment parameters are stored in a dictionary that's indexed by equations from the traced jaxpr. I think the only thing to really change is to index the dictionary by the stax layer rather than the jaxpreqn. As a bonus, it should be more straightforward to make the params dict more human readable. I'll have a go at implementing that so we can see what it looks like.

**Comment by levskaya** [comment]
Hey @j-towns you might want to check out an early version of some pointy name-binding constructs we've introduced at https://github.com/tensorflow/tensor2tensor/tree/master/tensor2tensor/trax/stax in slax.py - basically the Share, Bind, and Lambda decorators that enable simple pointiness in stax.  
I really love the idea of figuring out how to do all this natively inside JAX and leveraging the tracing machinery there to find the proper long-term solution to NN code, but the above hacks are pretty simple python tricks for letting stax scale a little further short-term.

**Comment by j-towns** [comment]
Another approach to doing 'implicit data-flow' in an uncompromisingly functional, JAX-ish, and lightweight way is to make functions into pytrees (with a flattener that returns all array constants/literals). This allows you to (for example) take gradients with respect to constants in a function. Very little code is required to do this.

You can bootstrap off of the fact that JAX already hoists array constants when tracing to a jaxpr. Below I define a `CallableJaxpr` class, which is just a `ClosedJaxpr` with a `__call__` method, and register the class as a pytree node so that gradients can be taken wrt it. For now the pytree node flattener just returns the jaxpr's consts but it could descend into the jaxpr looking for scalar literals.
```python
from functools import partial

from jax import tree_util
from jax import api_util
from jax.interpreters import partial_eval as pe
from jax import linear_util as lu
import jax.core as jc
from jax.abstract_arrays import raise_to_shaped

from jax.util import safe_map, safe_zip
from jax import random
from jax import nn
import jax.numpy as np
from jax import grad

map = safe_map
zip = safe_zip


class CallableJaxpr(jc.ClosedJaxpr):
    def __init__(self, jaxpr, consts, in_tree, out_tree):
        jc.ClosedJaxpr.__init__(self, jaxpr, consts)
        self.in_tree = in_tree
        self.out_tree = out_tree

    def __call__(self, *inputs):
        flat_inputs, in_tree = tree_util.tree_flatten(list(inputs))
        assert in_tree == self.in_tree
        flat_out = jc.eval_jaxpr(self.jaxpr, self.consts, *flat_inputs)
        return tree_util.tree_unflatten(self.out_tree, flat_out)

def flatten_callable_jaxpr(j):
    # Maybe we should descend into the jaxpr looking for literals.
    return j.consts, (j.jaxpr, j.in_tree, j.out_tree)

def unflatten_callable_jaxpr(meta, consts):
    jaxpr, in_tree, out_tree = meta
    return CallableJaxpr(jaxpr, consts, in_tree, out_tree)

tree_util.register_pytree_node(
    CallableJaxpr, flatten_callable_jaxpr, unflatten_callable_jaxpr)

def callable_jaxpr(f, example_inputs):
    """
    Wrap `fun` in a CallableJaxpr pytree node.
    """
    f = lu.wrap_init(f)
    flat_inputs, in_tree = tree_util.tree_flatten(list(example_inputs))
    flat_f, out_tree = api_util.flatten_fun_nokwargs(f, in_tree)
    in_avals = [raise_to_shaped(jc.get_aval(i)) for i in flat_inputs]
    jaxpr, out_avals, consts = pe.trace_to_jaxpr_dynamic(flat_f, in_avals)
    return CallableJaxpr(jaxpr, consts, in_tree, out_tree())


if __name__ == '__main__':
    rng = random.PRNGKey(0)
    rng, rng_weights, rng_biases, rng_x, rng_y = random.split(rng, 5)

    weights = random.normal(rng_weights, (3, 4))
    biases = random.normal(rng_biases, (3,))

    x = random.normal(rng_x, (4,))  # Example input
    y = random.normal(rng_y, (3,))  # Example label

    @partial(callable_jaxpr, example_inputs=[x])
    def nn(x):
        return nn.sigmoid(np.dot(weights, x) + biases)

    def loss(nn, x, y):
        return np.sum((nn(x) - y) ** 2)

    print(nn)
    # Take the gradient wrt the neural net
    print(grad(loss, 0)(nn, x, y))
```

**Comment by j-towns** [comment]
Another idea which I had today is to use generator functions, see https://gist.github.com/j-towns/026eaa335a043e5311fe34de97501319#file-randomize-py-L68-L102.
